articles/active-directory/develop/msal-net-token-cache-serialization.md

```diff
@@ -134,27 +134,35 @@ using Microsoft.Extensions.DependencyInjection;
 
 ```CSharp
 
- private static IConfidentialClientApplication app;
-
-public static async Task<IConfidentialClientApplication> BuildConfidentialClientApplication(
-  string clientId,
-  CertificateDescription certDescription,
-  string tenant)
+public static async Task<AuthenticationResult> GetTokenAsync(string clientId, X509Certificate cert, string authority, string[] scopes)
  {
-  if (app== null)
-  {
      // Create the confidential client application
      app= ConfidentialClientApplicationBuilder.Create(clientId)
        // Alternatively to the certificate you can use .WithClientSecret(clientSecret)
-       .WithCertificate(certDescription.Certificate)
+       .WithCertificate(cert)
        .WithLegacyCacheCompatibility(false)
-       .WithTenantId(tenant)
+       .WithAuthority(authority)
        .Build();
 
-     // Add an in-memory token cache. Other options available: see below
-     app.AddInMemoryTokenCache();
-   }
-   return app;
+     // Add a static in-memory token cache. Other options available: see below
+     app.AddInMemoryTokenCache();  // Microsoft.Identity.Web 1.16 or over
+   
+     // Make the call to get a token for client_credentials flow (app to app scenario) 
+     return await app.AcquireTokenForClient(scopes).ExecuteAsync();
+     
+     // OR Make the call to get a token for OBO (web api scenario)
+     return await app.AcquireTokenOnBehalfOf(scopes, userAssertion).ExecuteAsync();
+     
+     // OR Make the call to get a token via auth code (web app scenario)
+     try
+     {          
+          var account = await app.GetAccountsAsync(homeAccountId);
```

Comment:
No, you would never call AcquireTokenSilent in the auth code flow case, otherwise the authCode wouldn't be redeemed when you challenge the user (for instance when you ask them to do 2FA)

This should be only:

```CSharp
 return await app.AcquireTokenByAuthorizationCode(scopes, authCode);  
```

Reply:
Ack. So it would be expected that the session knows that a user has logged in and therefore a token is available in the cache.

Reply:
yes. In Web apps, the Http Context contains the claims that will help compute the account id